File: 24-06-2025_Lexicographically Largest String After K Deletions.java
class Solution {
    public static String maxSubseq(String s, int k) {
        int n = s.length();

        // Keep a copy of k to track deletions
        int toRemove = k;
        StringBuilder res = new StringBuilder();

        for (int i = 0; i < n; i++) {
            while (res.length() > 0 && toRemove > 0 &&
                   res.charAt(res.length() - 1) < s.charAt(i)) {
                res.deleteCharAt(res.length() - 1);
                toRemove--;
            }
            res.append(s.charAt(i));
        }

        // Result should be of length n - k
        return res.substring(0, n - k);
    }
}

File: 24-06-2025_Lexicographically Largest String After K Deletions.py
class Solution:

    def maxSubseq(self, s: str, k: int) -> str:
        n = len(s)
        res = ""
        # Keep a separate copy of k
        to_remove = k

        # Build the result greedily
        for i in range(n):
            while res and to_remove > 0 and res[-1] < s[i]:
                res = res[:-1]
                to_remove -= 1
            res += s[i]

        # Result should be of length n - k
        return res[:n - k]